src/coreComponents/constitutive/unitTests/testCompositionalProperties.cpp

```diff
@@ -0,0 +1,536 @@
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+// Source includes
+#include "codingUtilities/UnitTestUtilities.hpp"
+#include "constitutive/fluid/multifluid/compositional/functions/CompositionalProperties.hpp"
+#include "constitutive/fluid/multifluid/compositional/functions/CubicEOSPhaseModel.hpp"
+#include "TestFluid.hpp"
+
+namespace geos
+{
+namespace testing
+{
+
+static constexpr real64 relTol = 1.0e-4;
+static constexpr real64 absTol = 1.0e-7;
+
+template< integer NC >
+using CompositionalPropertiesTestData = std::tuple<
+  real64 const,     // 0 - pressure
+  real64 const,     // 1 - temperature
+  Feed< NC > const, // 2 - composition
+  real64 const,     // 3 - expected molar density
+  real64 const,     // 4 - expected molecular weight
+  real64 const      // 5 - expected mass density
+  >;
+
+template< typename EOS_TYPE, integer NC >
+std::vector< CompositionalPropertiesTestData< NC > > generateTestData();
+
+template< integer NC >
+std::unique_ptr< TestFluid< NC > > createFluid();
+template<>
+std::unique_ptr< TestFluid< 4 > > createFluid< 4 >()
+{
+  return TestFluid< 4 >::create( {Fluid::N2, Fluid::C8, Fluid::C10, Fluid::H2O} );
+}
+
+template< typename EOS_TYPE, integer NC >
+class CompositionalPropertiesTestDataTestFixture : public ::testing::TestWithParam< CompositionalPropertiesTestData< NC > >
+{
+public:
+  static constexpr integer numComps = NC;
+public:
+  CompositionalPropertiesTestDataTestFixture()
+    : m_fluid( createFluid< NC >() )
+  {}
+  ~CompositionalPropertiesTestDataTestFixture() = default;
+
+  // Compares the calculated molar density against the expected value from PVT package
+  void testMolarDensity( CompositionalPropertiesTestData< NC > const & data )
+  {
+    real64 const pressure = std::get< 0 >( data );
+    real64 const temperature = std::get< 1 >( data );
+    array1d< real64 > composition;
+    TestFluid< NC >::createArray( composition, std::get< 2 >( data ));
+    real64 const expectedMolarDensity = std::get< 3 >( data );
+
+    real64 molarDensity = 0.0;
+    computeMolarDensity( pressure,
+                         temperature,
+                         composition,
+                         molarDensity );
+    checkRelativeError( molarDensity, expectedMolarDensity, relTol, absTol );
+  }
+
+  // Compares the calculated molar density derivatives against numerical calculated
+  // finite difference values
+  void testMolarDensityDerivative( CompositionalPropertiesTestData< NC > const & data )
+  {
+    real64 const pressure = std::get< 0 >( data );
+    real64 const temperature = std::get< 1 >( data );
+    array1d< real64 > composition;
+    TestFluid< NC >::createArray( composition, std::get< 2 >( data ));
```

Comment:
Should we refactor that as it will be repeated in all tests to out the inputs and expected output ?